### DISCUSSION TITLE
Training TTS For a New Language

### DISCUSSION THREAD

**Discussion by skumarlabs** [coqui-ai/TTS#1198] [General Q&A]
So, I want to train the TTS for an Indian language Hindi. After going through the [docs](https://tts.readthedocs.io/en/latest/), I would say I have figured out few things which are needed to achieve the above. But now I think, the more I am reading `Discussions`, the more I am getting confused. Can someone please help me understand what exact steps are needed to train on a new model using my dataset. 

Below is what I have done based on my understanding so far:

1. I have prepared my dataset in `ljspeech` format so that my dataset have a `metadata.csv` and actual recording under `wavs` folder.
2. I have analyzed and cleaned my dataset as per guidlines at [formatting_your_dataset](https://tts.readthedocs.io/en/latest/formatting_your_dataset.html)
3. I have created a `config.json` file taking [tutorial_for_nervous_beginners](https://tts.readthedocs.io/en/latest/tutorial_for_nervous_beginners.html#cli-way) & [faqs](https://tts.readthedocs.io/en/latest/faq.html) as reference. which is below
```
{
    "model": "vits",
    "batch_size": 48,
    "eval_batch_size": 16,
    "num_loader_workers": 4,
    "num_eval_loader_workers": 4,
    "run_eval": true,
    "test_delay_epochs": -1,
    "epochs": 1000,
    "text_cleaner": "basic_cleaners",
    "use_phonemes": false,
    "phoneme_language": "hi",
    "phoneme_cache_path": "phoneme_cache",
    "print_step": 25,
    "print_eval": true,
    "mixed_precision": true,
    "max_seq_len":500000,
    "output_path": "TTS/recipes/ljspeech/vits_tts",
    "test_sentences": ["मेरे प्यारे देशवासियो", "आपके परिवार के सदस्य के रूप में उपस्थित हूँ।"],
    "datasets":[{"name": "ljspeech", "meta_file_train":"metadata.csv", "path": "/path/to/dataset/root"}],
    "characters":{
        "characters": "अआइईउऊऍएऐऑओऔकखगघचछजझटठडढणतथदधनपफबभमयरलवशषसह़ािीुूृेैॉोौ्ज़ड़ढ़फ़।–‘’“”!,-.?ँंः",
        "punctuations": "-–,.:;'‘’“”()!?ँंः।",
        "phonemes": null,
        "unique": true
    },
    "audio":{
        "sample_rate": 32000
    }
 }
```

Now, I tried following YourTTS notebooks as well but got stuck at `restoring model state dict` step because of below error:
`RuntimeError: Error(s) in loading state_dict for Vits:
	size mismatch for text_encoder.emb.weight: copying a param with shape torch.Size([165, 192]) from checkpoint, the shape in current model is torch.Size([103, 192]).`

Also, 

- I am not sure if I should retrain model from scratch or restore a pretrained model and finetune that since my language and it's character set is entirely different from that on which the pretrained models are available.

- Where does Speaker encoder model fits here in the process? Is it calculated automatically if I run the training with above config.json?

Can someone guide me here or point to a resource on what I need to do train a TTS on my Hindi dataset? Thanks for reading so far!

**Comment by jaggukaka** [comment]
I guess to get past the error may be you need to keep the original trained model's character set too. After that see how it goes. 
Speaker encoder is a separate model. 
There are two modes which you can use to have speaker specific embeddings, 

1. Use a pre-trained speaker encoder and generate d_vectors for all the samples you are including in your dataset 
2. Use speaker_embedding, which is another layer in the network and it learns the speaker specific embedding as you train the model.

But, probably the model which you are fine tuning is mostly trained with a speaker encoder, I guess it is better to generate d_vectors for every utterance in your data set and pass it to the model and not use speaker_embedding layer. (There are properties in config.json to set this appropriately)
You can generate d_vectors for your dataset by using this [script](https://github.com/coqui-ai/TTS/blob/main/TTS/bin/compute_embeddings.py).

Hope this information helps.

**Reply by skumarlabs** [reply]
Hey @jaggukaka ,

Thanks for your reply. Yes, I saw your other replies and clarification regarding `d_vectors` when going through discussions earlier.
So, now I understand that first I need to compute them. Then, I need to pass those vectors when using ` python train_tts.py --config_path config.json` 
One thing that still is not clear to me. Can I use pretrained vits model to continue fine tune on my dataset or I have to train a new model from scratch? since my dataset's character set is completely different from any other language on which the model is pretrained.

**Reply by jaggukaka** [reply]
If you have resources and time, then train from scratch, follow the yourtts paper, see experiments section, and try to follow same training method (for multiple languages of-course). This will probably give better results. 
Fine tuning existing model, I don't know if you add extra characters and additional language, the embeddings will restore correctly, I'm giving it a try, will let you know.
FYI, full training I haven't tried but just going by the merit of the paper.

**Reply by SaadBazaz** [reply]
How did it go for you guys? @jaggukaka @skumarlabs 
Were you able to build a model for Hindi?

**Reply by TejaswiniiB** [reply]
Hi @jaggukaka @skumarlabs ,
Were you able to build Hindi TTS ?

**Reply by skumarlabs** [reply]
No. Some of the utility scripts provided only work on linux OS. I didn't get a chance to look deeper.

**Reply by jaggukaka** [reply]
Sorry for late reply, it's been long since I worked on this, but yes I was successfully able to build for Telugu. Didn't try for hindi though, should be easy for hindi as there's more trading data out there.

**Reply by vikranthsagar** [reply]
Hi @jaggukaka did you train your model from scratch or did you use any of the pre-trained models?

**Reply by dsunjka** [reply]
@jaggukaka I would also be very interested in your lessons learned and any hints you might have for training for a new language.

**Reply by swapnil3597** [reply]
Hi @jaggukaka, can you please share how were you able to train Telugu TTS.

**Reply by Rajashekhar-Reddy** [reply]
@swapnil3597 Have you trained Telugu TTS successfully?

**Reply by Abhi84sia** [reply]
how to apply on hindi

**Reply by nagababumo** [reply]
> I guess to get past the error may be you need to keep the original trained model's character set too. After that see how it goes. Speaker encoder is a separate model. There are two modes which you can use to have speaker specific embeddings,
> 
> 1. Use a pre-trained speaker encoder and generate d_vectors for all the samples you are including in your dataset
> 2. Use speaker_embedding, which is another layer in the network and it learns the speaker specific embedding as you train the model.
> 
> But, probably the model which you are fine tuning is mostly trained with a speaker encoder, I guess it is better to generate d_vectors for every utterance in your data set and pass it to the model and not use speaker_embedding layer. (There are properties in config.json to set this appropriately) You can generate d_vectors for your dataset by using this [script](https://github.com/coqui-ai/TTS/blob/main/TTS/bin/compute_embeddings.py).
> 
> Hope this information helps.

hii can you help me in making on hindi please, it alot to me

**Comment by erogol** [comment]
Since your character set is totally different, you might need to train from scratch but restoring from the pre-trained model would at least help the vocoder to learn faster.

**Comment by Rajashekhar-Reddy** [comment]
@jaggukaka Have you tried to train Telugu?

**Reply by RAVINDRA8008** [reply]
Got any solution mate?

**Comment by souvikg544** [comment]
hello has anyone done it?

**Reply by Rajashekhar-Reddy** [reply]
Hi @souvikg544 , Have you started working on this?

**Reply by souvikg544** [reply]
Yes  ! Lets do together ...connect with me at - https://www.linkedin.com/in/souvikg544/

**Reply by souvikg544** [reply]
I am doing it on yourtts specifically

**Reply by saiakarsh193** [reply]
@souvikg544 hey even I am working on yourtts for Indian languages. I am facing a few difficulties. If possible can we connect on Linkedin and discuss further?

**Reply by VarunGumma** [reply]
Hi @souvikg544 @saiakarsh193 @Rajashekhar-Reddy, any update on this with YourTTS?

**Reply by sapienfrom2000s** [reply]
hey, any update on tts-hindi?

**Reply by saiakarsh193** [reply]
Hey, so I trained multiple Indian language TTS using Tacotron 2, Fast Pitch, VITS and YourTTS. The process is slightly different for each case but it is totally possible to train for a new language from scratch.

**Reply by sapienfrom2000s** [reply]
I need the VITS version for hindi and indian English if possible. Want to use piper (https://github.com/rhasspy/piper) with it.

**Reply by ianujkr** [reply]
did u trained on bengali language as well?

**Comment by varshkolla** [comment]
could you please let me know how to train the tts model for telugu. I have been trying to do using the AI4basic datset. @jaggukaka

**Reply by RAVINDRA8008** [reply]
Got any update? On telugu? Language?

**Comment by varshkolla** [comment]
Nope not yet
